// diffhtml/dist/cjs/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Internals = exports.html = exports.innerHTML = exports.outerHTML = exports.use = exports.createTree = exports.release = exports.removeTransitionState = exports.addTransitionState = exports.VERSION = undefined;

var _create = require('./tree/create');

var _create2 = _interopRequireDefault(_create);

var _create3 = require('./node/create');

var _create4 = _interopRequireDefault(_create3);

var _parseNewTree = require('./tasks/parse-new-tree');

var _parseNewTree2 = _interopRequireDefault(_parseNewTree);

var _reconcileTrees = require('./tasks/reconcile-trees');

var _reconcileTrees2 = _interopRequireDefault(_reconcileTrees);

var _internals = require('./util/internals');

var _internals2 = _interopRequireDefault(_internals);

var _parse = require('./util/parse');

var _parse2 = _interopRequireDefault(_parse);

var _innerHtml = require('./inner-html');

var _innerHtml2 = _interopRequireDefault(_innerHtml);

var _outerHtml = require('./outer-html');

var _outerHtml2 = _interopRequireDefault(_outerHtml);

var _transaction = require('./transaction');

var _html = require('./html');

var _html2 = _interopRequireDefault(_html);

var _release = require('./release');

var _release2 = _interopRequireDefault(_release);

var _use = require('./use');

var _use2 = _interopRequireDefault(_use);

var _transition = require('./transition');

var _version = require('./version');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// At startup inject the HTML parser into the default set of tasks.
_transaction.defaultTasks.splice(_transaction.defaultTasks.indexOf(_reconcileTrees2.default), 0, _parseNewTree2.default);

const api = {
  VERSION: _version.__VERSION__,
  addTransitionState: _transition.addTransitionState,
  removeTransitionState: _transition.removeTransitionState,
  release: _release2.default,
  createTree: _create2.default,
  use: _use2.default,
  outerHTML: _outerHtml2.default,
  innerHTML: _innerHtml2.default,
  html: _html2.default
};

const { assign } = Object;

// This is an internal API exported purely for middleware and extensions to
// leverage internal APIs that are not part of the public API. There are no
// promises that this will not break in the future. We will attempt to minimize
// changes and will supply fallbacks when APIs change.
//
// Note: The HTML parser is only available in this mode.
const Internals = assign(_internals2.default, api, {
  parse: _parse2.default,
  defaultTasks: _transaction.defaultTasks,
  tasks: _transaction.tasks,
  createNode: _create4.default,
  syncTree: _create2.default
});

// Attach a circular reference to `Internals` for ES/CJS builds.
api.Internals = Internals;

// Automatically hook up to DevTools if they are present.
if (typeof devTools !== 'undefined') {
  (0, _use2.default)(devTools(Internals));
  console.warn('diffHTML DevTools: Found and Activated...');
}

exports.VERSION = _version.__VERSION__;
exports.addTransitionState = _transition.addTransitionState;
exports.removeTransitionState = _transition.removeTransitionState;
exports.release = _release2.default;
exports.createTree = _create2.default;
exports.use = _use2.default;
exports.outerHTML = _outerHtml2.default;
exports.innerHTML = _innerHtml2.default;
exports.html = _html2.default;
exports.Internals = Internals;
exports.default = api;
